Content.Server/GameTicking/Rules/RevolutionaryRuleSystem.cs

```diff
@@ -0,0 +1,336 @@
+        for (var revs = 0; revs < headRevs; revs++)
+        {
+            IPlayerSession headRev;
+            if (prefList.Count == 0)
+            {
+                if (playerList.Count == 0)
+                {
+                    break;
+                }
+                headRev = _random.PickAndTake(playerList);
+            }
+            else
+            {
+                headRev = _random.PickAndTake(prefList);
+                playerList.Remove(headRev);
+            }
+            var mind = headRev.Data.ContentData()?.Mind;
+            if (mind == null)
+            {
+                continue;
+            }
+            _mindSystem.AddRole(mind, new RevolutionaryRole(mind, _prototypeManager.Index<AntagPrototype>(comp.HeadRevPrototypeId)));
+            var inCharacterName = string.Empty;
+            if (mind.OwnedEntity != null)
+            {
+                AddComp<RevolutionaryRuleComponent>(mind.OwnedEntity.Value);
+                AddComp<HeadRevolutionaryComponent>(mind.OwnedEntity.Value);
+                AddComp<RevolutionaryComponent>(mind.OwnedEntity.Value);
+                _stationSpawningSystem.EquipStartingGear(mind.OwnedEntity.Value, _prototypeManager.Index<StartingGearPrototype>(comp.HeadRevGearPrototypeId), null);
+                _npcFaction.RemoveFaction(mind.OwnedEntity.Value, "NanoTrasen", false);
+                _npcFaction.AddFaction(mind.OwnedEntity.Value, "Revolutionary");
+                inCharacterName = MetaData(mind.OwnedEntity.Value).EntityName;
+                comp.HeadRevs.Add(inCharacterName, headRev.Name);
+            }
+            if (mind.Session != null)
+            {
+                var message = Loc.GetString("head-rev-role-greeting");
+                var wrappedMessage = Loc.GetString("chat-manager-server-wrap-message", ("message", message));
+                _chatManager.ChatMessageToOne(Shared.Chat.ChatChannel.Server, message, wrappedMessage, default, false, mind.Session.ConnectedClient, Color.MediumBlue);
+            }
+        }
+        AssignCommandStaff(cantBeAntag);
+    }
+
+    //Assigns command staff based on their weight (real). I THINK all head weight is above 10 so uh let me know if not.
+    //I wanted to do sec as well but I'm not sure if they have a specific weight and they have glasses so if they get converted it's their own fault.
+    private void AssignCommandStaff(List<IPlayerSession> players)
+    {
+        foreach (var player in players)
+        {
+            var mind = player.GetMind();
+            if (mind != null && mind.CurrentJob != null)
+            {
+                if (mind.OwnedEntity != null && mind.CurrentJob.Prototype.Weight >= 10)
+                {
+                    AddComp<HeadComponent>(mind.OwnedEntity.Value);
+                }
+            }
+        }
+    }
+
+
+    private void OnMobStateChanged(MobStateChangedEvent ev)
+    {
+        if (ev.NewMobState == MobState.Dead)
+            CheckRevsLose(ev.Target);
+    }
```

Comment:
This shouldn't be using a broadcast subscription.

Reply:
Can you explain this one for me

Reply:
A broadcast subscription is the sort of event subscribed to indiscriminately, i.e. it will receive all events. When you subscribe to an event without specifying a component, you're making a broadcast subscription.

```
            SubscribeLocalEvent<ZombieComponent, MobStateChangedEvent>(OnMobState);
            SubscribeLocalEvent<MobStateChangedEvent>(OnMobStateChanged);
```

The first is a direct subscription. The second is a broadcast subscription.

A direct subscription will receive that event only for entities with the component specified.